### ISSUE TITLE
DependencyInjection Extentions crashes on Xamarin Android.

### ISSUE THREAD

**Issue by cwaldron** [dotnet/extensions#2355]
### Describe the bug
Using AddSingleton to setup IoC container, crashes on Android.  Using TryAddSingleton works.
However AddSignleton works as expected on the desktop.  The call to serviceCollection.BuildServiceProvider triggers the crash.

### To Reproduce
This is occurring in an application with a large number of interfaces, too many for me to reproduce.  I don't have the time to set up a small project with a few interfaces to see if I can repro the problem.  The best way to repro this issue is to create a Xamarin app and register interfaces using AddSingleton and another one using TryAddSingleton to see the difference.  Also, you must register your interfaces before resolving.  Trying to register after resolution cause the application to crash.

### Expected behavior
Expected the DI to work on Android device exactly as it would on the desktop.

### Additional context
Below is the stack track received from the debug window of the Xamarin Android app.
[bug.txt](https://github.com/aspnet/Extensions/files/3614483/bug.txt)

**Comment by davidfowl** [comment]
> This is occurring in an application with a large number of interfaces, too many for me to reproduce. I don't have the time to set up a small project with a few interfaces to see if I can repro the problem. The best way to repro this issue is to create a Xamarin app and register interfaces using AddSingleton and another one using TryAddSingleton to see the difference. Also, you must register your interfaces before resolving. Trying to register after resolution cause the application to crash.

How many interfaces and how deep is the graph?

**Comment by cwaldron** [comment]
There are 48 Interfaces.  Well actually 30 interfaces and 18 class types.  I've did my best to remove any cyclical definitions.  I don't think I'm encountering a cycle.  In fact I know I'm not because when I did hit a cycle I got a stack overflow.  Therefore I don't think that the graph is very deep.  Like I mentioned, the application works when using TryAddXXX but as soon as I use AddXXX it crashes however I can use AddXXX and even replace interfaces on the desktop but not when running on the device.

Cheers and thanks for the quick response.
